### DISCUSSION TITLE
How do I launch DAB at server start?

### DISCUSSION THREAD

**Discussion by GustavBrock** [Azure/data-api-builder#1885] [Q&A]
I have set up DAB to start and listen to port 8080 and 8081 using a command file like this, which I can call when logged on to the server as an admin user:

```
setlocal
set ASPNETCORE_URLS=http://api.example.com:8080;https://api.example.com:8081
"c:\program files\dab\dab.exe" start -c "c:\programdata\dab\dab-config.json" 
endlocal
```

But I wish this to happen whenever the server has rebooted - without a user login.
What is the recommended method to achieve that, please?

**Comment by severussundar** [comment]
Hi @GustavBrock, thank you for the question

One way to achieve this is by making use of the Windows Registry. You can create a script with the commands that you have provided in your description and create an entry in one of `HKEY_LOCAL_MACHINE_SOFTWARE\\Microsoft\\Windows\\CurrentVersion\\Run` or `HKEY_CURRENT_USER_SOFTWARE\\Microsoft\\Windows\\CurrentVersion\\Runonce` (There are more options available in the Windows registry, I would encourage you to kindly go through other options as well to determine the one that best fits your use-case).  Creating an entry in windows registry essentially enables the execution of your script as soon as the server starts/re-starts.

**Reply by GustavBrock** [reply]
Thanks a lot. Will check it out later today.

**Reply by GustavBrock** [reply]
@severussundar , that was not the best method, as Run and RunOnce are handled only at a user logon.

Ideally, it should run as a service, but I can't get sc.exe to create the service from dab.exe.
However, here is a poor-man's method taking advantage of _Task Scheduler_:

First, set the environment variable that specifies the listening ports in:
_Settings, System, About, Advance System Settings, System Properties_:

![image](https://github.com/Azure/data-api-builder/assets/6253835/53a04bc1-1173-41d1-b328-3214ed379575)

Click _Environment Variables_ and then _New ..._ and specify your URLs:

![image](https://github.com/Azure/data-api-builder/assets/6253835/6af4487a-b53f-46ee-af95-55ffd97c0650)

Click _OK_ to save it.

Next, open _Task Scheduler_ and create a new task, _DAB_ having properties like these where you use a user account having priviliges to execute _Dab_:

![image](https://github.com/Azure/data-api-builder/assets/6253835/84ddafb3-a9a3-43dd-92ac-a21e0b74c64c)

![image](https://github.com/Azure/data-api-builder/assets/6253835/88f7c907-a230-41f8-946a-ac3baa152605)

![image](https://github.com/Azure/data-api-builder/assets/6253835/40a1c378-0f98-47b5-879e-8c1d44643d2e)

![image](https://github.com/Azure/data-api-builder/assets/6253835/40f496dd-df44-449c-8d6c-98dbb179154b)

![image](https://github.com/Azure/data-api-builder/assets/6253835/6d433f92-8920-4c6c-b470-0683ee786ee0)

Save the task and try to run it manually; it should load, and the _Task Manager_ should display _dab.exe_ to be running:

![image](https://github.com/Azure/data-api-builder/assets/6253835/5f48881c-9d86-4d04-b3b7-40fc2f893a6e)

Anyone, feel free to suggest alternative methods.

**Comment by seantleonard** [comment]
There are instructions on how to host an ASP.NET app as a Windows Service here: https://learn.microsoft.com/en-us/aspnet/core/host-and-deploy/windows-service?view=aspnetcore-6.0&tabs=visual-studio

Additionally, you can host with IIS: https://learn.microsoft.com/en-us/aspnet/core/tutorials/publish-to-iis?view=aspnetcore-6.0&tabs=visual-studio

Note: these may require you to add additional settings to "appsettings.json" the is in DAB's package folder.

**Reply by GustavBrock** [reply]
But none of these methods can use dab.exe "as is", right? 
I would have to use the source code (not a big problem, I have VS2022 running).
